scripts/usb_scan.py

```diff
@@ -0,0 +1,545 @@
+
+    def get_all_storages(self):
+        ret = []
+        for i in self.interfaces:
+            ret.extend(i.get_storage_list())
+        return sorted(set(ret))
+
+    def is_child_of(self, parent):
+        # USB device is at the very top, since we don't care PCI info
+        return False
+
+
+class UsbInterface(UsbObject):
+    _NUMBER = "bInterfaceNumber"
+    _CLASS = "bInterfaceClass"
+    _SUB_CLASS = "bInterfaceSubClass"
+    _PROTOCOL = "bInterfaceProtocol"
+
+    _PROPS = [_NUMBER, _CLASS, _SUB_CLASS, _PROTOCOL]
+
+    def __init__(self, node, props):
+        super(UsbInterface, self).__init__(node)
+        for p in self._PROPS:
+            if p in props:
+                self[p] = props[p]
+        self.storages = set()
+
+    def __str__(self):
+        st = super(UsbInterface, self).__str__()
+        for s in self.storages:
+            st += "\n\t " + str(s)
+        return st
+
+    def is_class_hub(self):
+        return self._is_class_hub(self._CLASS)
+
+    def is_initialized(self):
+        for p in self._PROPS:
+            if p not in self:
+                return False
+        # TODO check a few important props(class, subclass, protocol) here
+        return super(self.__class__, self).is_initialized()
+
+    def get_storages(self):
+        return self.storages
+
+    def add_storage(self, storage):
+        if storage in self.storages:
+            Log().debug("overriding existing storage: " + storage)
+            self.storages.remove(storage)
+        self.storages.add(storage)
+
+    def del_storage(self, storage):
+        if storage in self.storages:
+            self.storages.remove(storage)
+        else:
+            Log().debug("try to remove unknown storage: " + storage)
+
+    # def get_device_node(self):
+    #     i_node = self.get_node()
+    #     i_colon = i_node.find(":")
+    #     if i_colon > 0:
+    #         return i_node[0:i_colon]
+
+    def is_child_of(self, parent):
+        if type(parent) is UsbDevice and parent.is_initialized():
+            conf_value = parent[UsbDevice._CONF_VALUE]
+            pattern = "^" + re.escape(parent.get_node()) + ":" + conf_value + "\.\d+"
+            return re.match(pattern, self.get_node())
+        return False
+
+
+class UsbStorage(UsbObject):
+    _PATH = "path"
+    _PROPS = [_PATH]
+
+    def __init__(self, node, path):
+        super(UsbStorage, self).__init__(node)
+        self[self._PATH] = path
+
+    def is_class_hub(self):
+        return False
+
+    def is_initialized(self):
+        path = self.get(self._PATH)
+        if path is None or path.find("usb") == -1:
+            return False
+        return super(self.__class__, self).is_initialized()
+
+    def get_parent_node(self):
+        parts = self[self._PATH].split("/")
+        usb = False
+        for part in parts:
+            if usb:
+                if part.find(":") != -1:
+                    return part
+            elif part.startswith("usb"):
+                usb = True
+
```

Comment:
What should be returned if this line is reached?

Reply:
it should return None.
And None is checked in is_child_of.

Reply:
explicit return None added.